infer/src/base/DBWriter.ml

```diff
@@ -363,22 +363,43 @@ module Server = struct
 
   let socket_exists () = in_results_dir ~f:(fun () -> Sys.file_exists_exn socket_name)
 
-  let server () =
-    L.debug Analysis Quiet "Sqlite write daemon: starting up@." ;
+  (* Error recuperation is done by attempting this function at module initialization time, and
+     not using DbWriter at all in case it fails. See {!can_use_socket} below. *)
+  let setup_socket () =
     if socket_exists () then L.die InternalError "Sqlite write daemon: socket already exists@." ;
     let socket = Unix.socket ~domain:socket_domain ~kind:Unix.SOCK_STREAM ~protocol:0 () in
     in_results_dir ~f:(fun () -> Unix.bind socket ~addr:socket_addr) ;
     (* [backlog] is (supposedly) the length of the queue for pending connections ;
        there are no rules about the implied behaviour though.  Here use optimistically
        the number of workers, though even that is a guess. *)
     Unix.listen socket ~backlog:Config.jobs ;
+    socket
+
+
+  let remove_socket socket =
+    in_results_dir ~f:(fun () ->
+        Unix.close socket ;
+        Unix.unlink socket_name )
+
+
+  (* Check whether we can create a socket to communicate with the asynchronous DBWriter process. *)
+  let can_use_socket () =
+    try
+      (* This function is called very early, and the infer-out directory may not have been created
+         yet. Ensure it exists before attempting to create the socket. *)
+      Unix.mkdir_p Config.toplevel_results_dir ;
```

Comment:
That's a bit too disgusting :) Could we have `use_daemon` be a lazy value instead?

Reply:
Good point, I suspect so. Will experiment and report.

Reply:
It works \o/.